text/0000-stabilize-arbitrary-self-types.md

```diff
@@ -320,22 +460,47 @@ Please also take into consideration that rust sometimes intentionally diverges f
     ```
     This fails, because `T: Receiver<Target=T>` generally does not hold.
     An alternative would be to lift the associated type into a generic type parameter of the `Receiver` trait, that would allow adding a blanket `impl Receiver<T> for T` without overlap.
-
-- Should this work for (statically resolved) trait method calls? e.g.
-  ```rs
-  use std::ops::Receiver;
-  struct CustomPtr<T>(T);
-  impl<T> Receiver for CustomPtr<T> {
-    type Target = T;
-  }
-  trait Foo {
-     fn bar(self: CustomPtr<Self>);
-  }
-  struct Baz;
-  impl Foo for Baz {
-    fn bar(self: CustomPtr<Self>) {}
-  }
-  ```
+- This sinister TODO is present in the code:
+                // FIXME(arbitrary_self_types): We probably should limit the
+                // situations where this can occur by adding additional restrictions
+                // to the feature, like the self type can't reference method substs.
```

Comment:
It's short for "substitutions".... but beyond that I have no idea :) We need to figure this out.

Here's my limited understanding here:
* The code in `wfcheck.rs` tries to identify realistic method candidates. It nearly always does a good job.
* Sometimes, it gets it wrong, and then this later code reports an error (at the place where this `FIXME` is)
* This seems to be the case where there are `self` types which implement `Receiver` but otherwise are a bit crazy, so the code in `wfcheck.rs` gets it wrong.
* If I add a `panic!` in place of this `FIXME` then there's one test in the rustc test suite which panics, so we have a small snippet of code showing what goes wrong. It's this one: https://github.com/rust-lang/rust/blob/master/tests/ui/self/arbitrary-self-from-method-substs.rs
* but when I adjust this test from being `Deref` to being `Receiver` it seems mostly to work.

Reply:
The Fixme is from here: https://github.com/rust-lang/rust/blob/f0580df0d53d67ad5c7f85756eb9f221566e4fb0/compiler/rustc_hir_typeck/src/method/confirm.rs#L511-L513

Some maybe relevant info from [compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs](https://github.com/rust-lang/rust/blob/f0580df0d53d67ad5c7f85756eb9f221566e4fb0/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs#L1390-L1402):
- "Substitute the values for the type parameters into the type of the referenced item."
- "In the case of `Foo<T>::method` and `<Foo<T>>::method`, if `method` is inherent, there is no `Self` parameter; instead, the impl needs type parameters, which we can infer by unifying the provided `Self` with the substituted impl type. This also occurs for an enum variant on a type alias."